File: apps/multiplatform/common/src/commonMain/kotlin/chat/simplex/common/views/usersettings/GiveNameView.kt
package chat.simplex.common.views.usersettings

import SectionBottomSpacer
import SectionItemView
import SectionTextFooter
import SectionView
import androidx.compose.foundation.layout.*
import androidx.compose.material.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import chat.simplex.common.model.*
import chat.simplex.common.platform.*
import chat.simplex.common.ui.theme.*
import chat.simplex.common.views.helpers.*
import chat.simplex.res.MR
import dev.icerock.moko.resources.compose.stringResource

// Giving a name to a contact.
//
// Only contacts whose profile carries a receiving address can be chosen: the
// name is sent to a one-time address derived from it, which is what keeps the
// transfer from linking the recipient to the name for everyone watching. A
// contact on an older version, or one who has not turned receiving on, simply
// has nowhere for it to land.
@Composable
fun GiveNameView(rhId: Long?, fqdn: String, close: () -> Unit) {
  val busy = remember { mutableStateOf(false) }
  val label = fqdn.removeSuffix(".simplex")

  val candidates = chatModel.chats.value.mapNotNull { c ->
    val ci = c.chatInfo
    if (ci is ChatInfo.Direct) ci.contact else null
  }
  val canReceive = candidates.filter { it.profile.metaAddress != null }
  val cannotReceive = candidates.size - canReceive.size

  ColumnWithScrollBar {
    AppBarTitle(stringResource(MR.strings.names_give_title))
    SectionTextFooter(stringResource(MR.strings.names_give_intro).format(fqdn))

    SectionView(stringResource(MR.strings.names_give_choose).uppercase()) {
      if (canReceive.isEmpty()) {
        SectionItemView { Text(stringResource(MR.strings.names_give_nobody), color = MaterialTheme.colors.secondary) }
      } else {
        canReceive.forEach { ct ->
          SectionItemView(click = {
            if (!busy.value) confirmGive(rhId, label, fqdn, ct, busy) { close() }
          }, disabled = busy.value) {
            Text(ct.profile.profileViewName)
          }
        }
      }
    }
    SectionTextFooter(
      if (cannotReceive == 1) stringResource(MR.strings.names_give_one_cannot)
      else if (cannotReceive > 0) stringResource(MR.strings.names_give_some_cannot).format(cannotReceive)
      else stringResource(MR.strings.names_give_footer)
    )
    SectionBottomSpacer()
  }
}

private fun confirmGive(rhId: Long?, label: String, fqdn: String, ct: Contact, busy: MutableState<Boolean>, onDone: () -> Unit) {
  AlertManager.shared.showAlertDialog(
    title = generalGetString(MR.strings.names_give_confirm_title),
    // Irreversible without their cooperation: after this they own it, not you.
    text = generalGetString(MR.strings.names_give_confirm_text).format(fqdn, ct.profile.profileViewName),
    confirmText = generalGetString(MR.strings.names_give_action),
    destructive = true,
    onConfirm = {
      withBGApi {
        busy.value = true
        try {
        // localDisplayName, not profile.displayName: the core resolves this
        // against local_display_name, which is deduplicated. Two contacts
        // called "alice" would otherwise send the name to whichever one the
        // core happened to match, irreversibly.
        val gifted = chatModel.controller.apiNameGift(rhId, label, "@${ct.localDisplayName}")
        if (gifted != null) {
          // The core sends the transfer message itself, so every client
          // behaves the same and this does not send a second one.
          AlertManager.shared.showAlertMsg(
            title = generalGetString(MR.strings.names_give_sent_title),
            text = generalGetString(MR.strings.names_give_sent_text).format(fqdn, ct.profile.profileViewName),
          )
          onDone()
        }
        } finally { busy.value = false }
      }
    }
  )
}

File: apps/ios/Shared/Views/UserSettings/SimplexNames/GiveNameView.swift
import SwiftUI
import SimpleXChat

// Giving a name to a contact. Only contacts whose profile carries a receiving
// address can be chosen; the name is sent to a one-time address derived from it.
struct GiveNameView: View {
    @Environment(\.dismiss) var dismiss
    let fqdn: String
    private var label: String { fqdn.replacingOccurrences(of: ".simplex", with: "") }

    private var candidates: [Contact] {
        ChatModel.shared.chats.compactMap { if case let .direct(c) = $0.chatInfo { return c } else { return nil } }
    }
    private var canReceive: [Contact] { candidates.filter { $0.profile.metaAddress != nil } }
    private var cannotReceive: Int { candidates.count - canReceive.count }

    var body: some View {
        List {
            Section("Give to") {
                if canReceive.isEmpty {
                    Text("None of your contacts can receive a name yet.").foregroundColor(.secondary)
                } else {
                    ForEach(canReceive) { ct in
                        Button(ct.profile.profileViewName) { confirm(ct) }
                    }
                }
            } footer: {
                if cannotReceive == 1 { Text("One of your contacts cannot receive names yet — they need a newer app version, or to turn receiving on.") }
                else if cannotReceive > 0 { Text(String.localizedStringWithFormat(NSLocalizedString("%d of your contacts cannot receive names yet — they need a newer app version, or to turn receiving on.", comment: ""), cannotReceive)) }
                else { Text("Only contacts who can receive names are listed.") }
            }
        }
        .navigationTitle("Give this name away")
    }

    private func confirm(_ ct: Contact) {
        showAlert(NSLocalizedString("Give the name away?", comment: "alert"),
                  message: String.localizedStringWithFormat(NSLocalizedString("%1$@ will belong to %2$@. Only they will be able to change or move it after this.", comment: "alert"), fqdn, ct.profile.profileViewName),
                  actions: {[
                    UIAlertAction(title: NSLocalizedString("Give it away", comment: ""), style: .destructive) { _ in
                        Task {
                            if await apiNameGift(label, recipient: "@\(ct.localDisplayName)") {
                                if let u = try? await apiGetActiveUser() { await MainActor.run { ChatModel.shared.updateUser(u) } }
                                await MainActor.run {
                                    showAlert(NSLocalizedString("Name given away", comment: "alert"), message: String.localizedStringWithFormat(NSLocalizedString("%1$@ now belongs to %2$@. It appears under names given to them.", comment: "alert"), fqdn, ct.profile.profileViewName))
                                    dismiss()
                                }
                            }
                        }
                    }, cancelAlertAction
                  ]})
    }
}
